client/go/dogma/content_service.go

```diff
@@ -0,0 +1,306 @@
+	if len(jsonPaths) != 0 {
+		if query, err = getJSONPaths(path, jsonPaths); err != nil {
+			return "", err
+		}
+
+		if len(revision) != 0 {
+			// have both of the jsonpath and the revision
+			query += "&revision=" + revision
+		}
+	} else if len(revision) != 0 {
+		// have the revision only
+		query += "?revision=" + revision
+	}
+	return query, nil
+}
+
+func getJSONPaths(path string, jsonPaths []string) (query string, err error) {
+	if !strings.HasSuffix(strings.ToLower(path), "json") {
+		return "", fmt.Errorf("the extension of the file should be .json (path: %v)", path)
+	}
+	for i, jsonPath := range jsonPaths {
+		if i == 0 {
+			query = "?"
+		} else {
+			query += "&"
+		}
+		query += fmt.Sprintf("jsonpath=%v", jsonPath)
+	}
+	// the query starts with '?'
+	return query, nil
+}
+
+func (con *contentService) getFiles(ctx context.Context,
+	projectName, repoName, revision, pathPattern string) ([]*Entry, *http.Response, error) {
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/contents%v", DefaultPathPrefix, projectName, repoName, pathPattern)
+
+	if len(revision) != 0 {
+		u += "?revision=" + revision
+	}
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var entries []*Entry
+	res, err := con.client.do(ctx, req, &entries)
+	if err != nil {
+		return nil, res, err
+	}
+	return entries, res, nil
+}
+
+func (con *contentService) getHistory(ctx context.Context,
+	projectName, repoName, from, to, pathPattern string) ([]*Commit, *http.Response, error) {
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/commits/%v", DefaultPathPrefix, projectName, repoName, from)
+
+	query := ""
+	if len(pathPattern) != 0 {
+		query = fmt.Sprintf("?path=%v", pathPattern)
+		if len(to) != 0 {
+			query += fmt.Sprintf("&to=%v", to)
+		}
+	} else if len(to) != 0 {
+		query = fmt.Sprintf("?to=%v", to)
+	}
+	u += query
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var commits []*Commit
+	res, err := con.client.do(ctx, req, &commits)
+	if err != nil {
+		return nil, res, err
+	}
+	return commits, res, nil
+}
+
+func (con *contentService) getDiff(ctx context.Context,
+	projectName, repoName, from, to, path string, jsonPaths []string) (*Change, *http.Response, error) {
+	if len(path) == 0 {
+		return nil, nil, fmt.Errorf("the path should not be empty")
+	}
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	u := fmt.Sprintf("%vprojects/%v/repos/%v/compare", DefaultPathPrefix, projectName, repoName)
+	query, err := getDiffQuery(from, to, path, jsonPaths)
+	if err != nil {
+		return nil, nil, err
+	}
+	u += query
+
+	req, err := con.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	change := new(Change)
+	res, err := con.client.do(ctx, req, change)
+	if err != nil {
+		return nil, res, err
+	}
+
+	return change, res, nil
+}
+
+func getDiffQuery(from, to, path string, jsonPaths []string) (query string, err error) {
```

Comment:
Ditto - How do we support other query languages than JSON path? e.g. http://jmespath.org/